### PR TITLE
Add eth_sendTransaction

### LINKED ISSUE #764: Eth RPC for eth_sendTransaction
Implement the eth_sendTransaction RPC to sign and encode transactions without the execution layer needing the secp256k1 private key.

note: Web3Signer will need to have support for loading secp256k1 keys and signing them. This functionality already exists due to filecoin signing in Web3Signer.

**We don't need to implement go-quorum private transaction support only EEA transactions. So the privateFor/go-quorum private transaction code should be removed as part of the migration.**

https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_sendtransaction

Tasks
- [x] Add configuration for chainId, downstream web3 provider
- [x] Add eth_sendTransaction to the Vertx routers for the Eth1Runner
- [x] Implement the eth_sendTransaction handler. Code for decoding, signing and sending the eth_sendRawTransaction can be adapted from [sendTransaction](https://github.com/ConsenSys/ethsigner/tree/master/ethsigner/core/src/main/java/tech/pegasys/ethsigner/core/requesthandler/sendtransaction) implementation in EthSigner.
- [x] Implement the nonce lookup using eth_getTransactionCount. Can be adapted from the EthSigner code.
- [x] Load the secp keys using the config in the key directory (The majority of this should be in place from the filecoin implementation)
- [ ] Add eea_sendTransaction  
- [ ] Integration tests 
- [ ] Acceptance tests

### PR DESCRIPTION
**Pull request by gfukushima**
## PR Description
Add eth_sendTransaction RPC method under the eth1 subcommand.

## Fixed Issue(s)

Fixes #764 
## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

## Testing

- [ ] I thought about testing these changes in a realistic/non-local environment.

### CODE REVIEW

File: core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/sendtransaction/transaction/EthTransactionTest.java
Comment (siladu) [inline core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/sendtransaction/transaction/EthTransactionTest.java:1]: This appears to be a duplicate of https://github.com/Consensys/web3signer/blob/master/core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/transaction/EthTransactionTest.java

I think the one in this PR should be deleted, because the one already in master matches up with this package structure:
https://github.com/Consensys/web3signer/blob/master/core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/sendtransaction/transaction/EthTransaction.java

Comment (siladu) [reply core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/sendtransaction/transaction/EthTransactionTest.java:1]: There's actually a number of duplicates, some probably valid...

```
find . -name "*.java" -exec basename {} \; | sort | uniq -d

ByteUtilsTest.java
CertificateHelpers.java
DatabaseUtil.java
DownstreamPathCalculatorTest.java
EthTransactionTest.java
IntegrationTestBase.java
PrivateTransaction.java
SignedAttestation.java
SignedBlock.java
Signer.java
TlsOptions.java
Transaction.java


find . -name "*.java" -exec basename {} \; | sort | uniq -d | xargs -I {} find . -name {}

./signing/src/test/java/tech/pegasys/web3signer/signing/util/ByteUtilsTest.java
./core/src/test/java/tech/pegasys/web3signer/core/util/ByteUtilsTest.java
./acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/tls/support/CertificateHelpers.java
./keystorage/src/testFixtures/java/tech/pegasys/web3signer/keystore/hashicorp/dsl/certificates/CertificateHelpers.java
./acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/utils/DatabaseUtil.java
./slashing-protection/src/testFixtures/java/db/DatabaseUtil.java
./core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/sendtransaction/DownstreamPathCalculatorTest.java
./core/src/test/java/tech/pegasys/web3signer/core/service/DownstreamPathCalculatorTest.java
./core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/sendtransaction/transaction/EthTransactionTest.java
./core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/transaction/EthTransactionTest.java
./core/src/integrationTest/java/tech/pegasys/web3signer/core/jsonrpcproxy/IntegrationTestBase.java
./slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/IntegrationTestBase.java
./core/src/integrationTest/java/tech/pegasys/web3signer/core/jsonrpcproxy/model/jsonrpc/PrivateTransaction.java
./core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/sendtransaction/transaction/PrivateTransaction.java
./slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedAttestation.java
./slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/model/SignedAttestation.java
./slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedBlock.java
./slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/model/SignedBlock.java
./acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/signer/Signer.java
./signing/src/main/java/tech/pegasys/web3signer/signing/secp256k1/Signer.java
./keystorage/src/main/java/tech/pegasys/web3signer/keystorage/hashicorp/config/TlsOptions.java
./core/src/main/java/tech/pegasys/web3signer/core/config/TlsOptions.java
./core/src/integrationTest/java/tech/pegasys/web3signer/core/jsonrpcproxy/model/jsonrpc/Transaction.java
./core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/sendtransaction/transaction/Transaction.java
```

Comment (gfukushima) [reply core/src/test/java/tech/pegasys/web3signer/core/service/jsonrpc/sendtransaction/transaction/EthTransactionTest.java:1]: Base on this list I've removed and identified the following: 

ByteUtilsTest.java - Deleted
DownstreamPathCalculatorTest.java - Deleted
EthTransactionTest.java - Deleted
TlsOptions.java - Interface and class with same name
Transaction.java - Interface and class with same name

I've only addressed the ones related to this PR

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/besu/BesuNode.java
Comment (siladu) [inline acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/besu/BesuNode.java:59]: It's used in this file so think we could remove this?

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/eth1rpc/Eth1RpcReloadKeysTest.java
Comment (siladu) [inline acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/eth1rpc/Eth1RpcReloadKeysTest.java:87]: nit: could be private

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/publickeys/KeyIdentifiersAcceptanceTestBase.java
Comment (siladu) [inline acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/publickeys/KeyIdentifiersAcceptanceTestBase.java:158]: Did we discuss this already on another PR? Maybe it's OK to always set the chainId? Not sure.

File: core/src/integrationTest/java/tech/pegasys/web3signer/core/jsonrpcproxy/DefaultTestBase.java
Comment (siladu) [inline core/src/integrationTest/java/tech/pegasys/web3signer/core/jsonrpcproxy/DefaultTestBase.java:17]: nit: If this can't be merged into IntegrationTestBase then would prefer a more obvious name such as ChainIdTestBase or Eth1TestBase

File: core/src/main/java/tech/pegasys/web3signer/core/Eth1Runner.java
Comment (siladu) [inline core/src/main/java/tech/pegasys/web3signer/core/Eth1Runner.java:105]: nit: if the order matters maybe can express that in the type by using ArrayList explicitly (and maybe extracting a variable with a name that mentions 'ordered')

File: core/src/main/java/tech/pegasys/web3signer/core/Runner.java
Comment (siladu) [inline core/src/main/java/tech/pegasys/web3signer/core/Runner.java:241]: nit: As per other comment, think this could be explicitly ArrayList to express the ordered property and name should change to reflect it

```suggestion
      final ArrayList<ArtifactSignerProvider> orderedArtifactSignerProviders,
```

File: core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java
Comment (siladu) [inline core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java:29]:
```suggestion
  ArrayList<ArtifactSignerProvider> orderedArtifactSignerProviders;

  public ReloadHandler(ArrayList<ArtifactSignerProvider> orderedArtifactSignerProviders) {
    this.orderedArtifactSignerProviders = orderedArtifactSignerProviders;
  }
```

File: core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java
Comment (siladu) [inline core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java:37]:
```suggestion
                orderedArtifactSignerProviders.stream()
```

**Review by siladu — COMMENTED** [review]
LGTM, just some nits. Won't approve yet as would like a second opinion especially about the loader changes. 

I reviewed the changed files. Mostly skipped over the obviously copy/pasted files but reviewed the file structure/package layout.

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/publickeys/KeyIdentifiersAcceptanceTestBase.java
Comment (gfukushima) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/publickeys/KeyIdentifiersAcceptanceTestBase.java]: removed, think it crawled back somehow when I reverted my previous PR

File: core/src/integrationTest/java/tech/pegasys/web3signer/core/jsonrpcproxy/DefaultTestBase.java
Comment (gfukushima) [reply core/src/integrationTest/java/tech/pegasys/web3signer/core/jsonrpcproxy/DefaultTestBase.java]: Merged

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/besu/BesuNode.java
Comment (gfukushima) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/besu/BesuNode.java]: removed

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/eth1rpc/Eth1RpcReloadKeysTest.java
Comment (gfukushima) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/tests/eth1rpc/Eth1RpcReloadKeysTest.java]: done

File: core/src/main/java/tech/pegasys/web3signer/core/Eth1Runner.java
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/Eth1Runner.java:105]: Lists are by definition ordered. Also rather not tie this to a specific implementation of List as it doesn't matter for the reloadHandler

File: core/src/main/java/tech/pegasys/web3signer/core/Runner.java
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/Runner.java]: This should stay as a type of `List` instead of changing to an `ArrayList`. As the `List` type is by definition ordered. Also rather not tie this to a specific implementation of List as it doesn't matter for the reloadHandler

Comment (siladu) [reply core/src/main/java/tech/pegasys/web3signer/core/Runner.java]: I was wrong, List is indeed ordered, I must have been thinking of Collection :)

File: signing/src/main/java/tech/pegasys/web3signer/signing/secp256k1/SingleSignerProvider.java
Comment (jframe) [inline signing/src/main/java/tech/pegasys/web3signer/signing/secp256k1/SingleSignerProvider.java:21]: This should be moved to the integration test package as this is only needed for the integration test now, we don't have a single provider mode in web3signer.

File: acceptance-tests/src/test/resources/besu/eth_hash_2018_no_replay_protection.json
Comment (siladu) [reply acceptance-tests/src/test/resources/besu/eth_hash_2018_no_replay_protection.json:6]: My PR adds londonBlock to one of these files at least

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java
Comment (gfukushima) [reply signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java:54]: In what scenario would that be possible? Specially if we consider that we're looping through the signers we've just loaded to the memory. As far as I remember we discussed that no signers are removed in run time, only added.

File: core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java:43]: 200 as we just submit the thread executor and return 200. Similar behaviour to what we had previously with the asycn call of the future if I'm not mistaken.

File: core/src/main/java/tech/pegasys/web3signer/core/Eth1Runner.java
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/Eth1Runner.java]: done

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java
Comment (gfukushima) [reply signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java]: changed

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java
Comment (gfukushima) [reply signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java:47]: As we spoke I'd rather keep them separate for now as grouping call of the "provided signerProvider" with the signer provider form the decorator has resulted in some other collateral effects like we would need some changes to identify what we want to return to the called of the methods. For instance if something calls avalilableIdentifiers() what should be returned public keys or eth1 addresses? a

File: signing/src/main/java/tech/pegasys/web3signer/signing/secp256k1/SingleSignerProvider.java
Comment (gfukushima) [reply signing/src/main/java/tech/pegasys/web3signer/signing/secp256k1/SingleSignerProvider.java]: moved

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java
Comment (siladu) [inline signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java:66]: nit: coming back to this code, it's not clear what this `String identifier` represents - I had to follow from EthSignTransactionResultProvider down to here and I guess it's an ethereum address based on the load method above. Maybe a better name at the RPC level would help.
Separately, EthSecpArtifactSigner uses a public key instead which makes this more confusing. 
I think it's probably a wider issue of using a String type in the interface so maybe out of scope for this PR.

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/sendtransaction/SendTransactionHandler.java
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/sendtransaction/SendTransactionHandler.java:81]: done

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/sendtransaction/transaction/VertxStoreRawRequestTransmitter.java
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/sendtransaction/transaction/VertxStoreRawRequestTransmitter.java:41]: removed

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java
Comment (gfukushima) [reply signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdpater.java:33]: applied

File: core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/signing/TransactionSerializer.java
Comment (gfukushima) [reply core/src/main/java/tech/pegasys/web3signer/core/service/jsonrpc/handlers/signing/TransactionSerializer.java:39]: Doing this would mean creating a new SignerForIdentifier with the signerProviderAdapted just to avoid the cast which is done the same way but in a more elegant way using the generics in the SignerForIdentifier class in the formatSignature method. I see this cast as a safe cast but happy to switch if we're strongly against the cast.

File: keystorage/src/integrationTest/java/tech/pegasys/web3signer/keystore/hashicorp/HashicorpIntegrationTest.java
Comment (jframe) [inline keystorage/src/integrationTest/java/tech/pegasys/web3signer/keystore/hashicorp/HashicorpIntegrationTest.java:73]: Why is this test disabled?

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java
Comment (jframe) [inline signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java:58]: nit: this would be a bit easier to read if this was extracted to own method

File: keystorage/src/integrationTest/java/tech/pegasys/web3signer/keystore/hashicorp/HashicorpIntegrationTest.java
Comment (jframe) [inline keystorage/src/integrationTest/java/tech/pegasys/web3signer/keystore/hashicorp/HashicorpIntegrationTest.java:73]: Why is this test disabled?

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java
Comment (jframe) [inline signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java:44]: Worth adding tests for the adapter?

File: core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/ReloadHandler.java:43]: Think will return a 500, which is the correct thing to do

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java
Comment (siladu) [inline signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java:58]: nit: eth1Address may be even better

Comment (gfukushima) [reply signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java:58]: my first change was eth1Address but then I thought that eth1 would be sort of implicit when it comes to secp, I could be wrong though

File: signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java
Comment (jframe) [inline signing/src/main/java/tech/pegasys/web3signer/signing/config/SecpArtifactSignerProviderAdapter.java:51]: nit: could use a method reference here

**Comment by jframe** [comment]
Can you also add a changelog entry for this

**Review by siladu — APPROVED** [review]
LGTM

**Pull request merged**
